source/s3_client.c

```diff
@@ -1909,7 +1909,10 @@ void s_acquire_mem_and_prepare_request(
         }
         /* END CRITICAL SECTION */
 
-        aws_future_s3_buffer_ticket_register_callback(payload->buffer_future, s_on_pool_buffer_reserved, payload);
+        /* Note: run callback async on event loop. this is done to avoid any race conditions between cancelling
+           which requires a lock on meta request and buffer acquire callback which also requires a lock. */
+        aws_future_s3_buffer_ticket_register_event_loop_callback(
```

Comment:
looks like when the future fails, `s_on_pool_buffer_reserved` will not remove the node from the list.

Reply:
fixed and added test